### PR TITLE
New CLI Flag `--devel` To Include Development/Prerelease Versions of Charts

### LINKED ISSUE #94: [Feature request] Repository tab support to  display helm that only have pre-release versions
we have helm that version is pre-release and only have pre-release version, while dashboard default search stable version, so I can not find my helm chart in the Repository tab.

Can helm-dashboard`s Repository tab support to  display helm that only have pre-release versions ?

just add `--devel` flag in `helm search repo` command can get the pre-release version.

ref: https://helm.sh/docs/helm/helm_search_repo/

```
# Search for release versions matching the keyword "nginx", including pre-release versions
$ helm search repo nginx --devel
```

edit

  https://github.com/komodorio/helm-dashboard/blob/cf407c63a23e25cd15878661c2067458a14979b3/pkg/dashboard/subproc/data.go#L191   
, like this 
```
func (d *DataLayer) ChartRepoVersions(chartName string) (res []*RepoChartElement, err error) {
	search := "/" + chartName + "\v"
	if strings.Contains(chartName, "/") {
		search = "\v" + chartName + "\v"
	}

	cmd := []string{"search", "repo", "--devel", "--regexp", search, "--versions", "--output", "json"}
	out, err := d.runCommandHelm(cmd...)
	if err != nil {
		return nil, err
	}

	err = json.Unmarshal([]byte(out), &res)
	if err != nil {
		return nil, err
	}
	return res, nil
}

func (d *DataLayer) ChartRepoCharts(repoName string) (res []*RepoChartElement, err error) {
	cmd := []string{"search", "repo", "--devel", "--regexp", "\v" + repoName + "/", "--output", "json"}
	out, err := d.runCommandHelm(cmd...)
	if err != nil {
		return nil, err
	}

	err = json.Unmarshal([]byte(out), &res)
	if err != nil {
		return nil, err
	}

	ins, err := d.ListInstalled()
	if err != nil {
		return nil, err
	}

	enrichRepoChartsWithInstalled(res, ins)

	return res, nil
}
```
or  give a args to control this feature.

### PR DESCRIPTION
**Pull request by Bhargav-InfraCloud**
# Motivation
[Issue][1] #94 

# Changes
1. New CLI flag `--devel`.
2. Tests to cover `Charts()`  and `Containing()`.

# Summary
CLI flag `--devel` to list/include development (prerelease versions like release candidate, alpha, beta, etc.) in the following endpoints:
- /api/helm/repositories/\<repo>
- /api/helm/repositories/versions?name=\<repo>
- /api/helm/repositories/latestver?name=\<repo>

# Related Issues
Closes #94 

[1]: https://github.com/komodorio/helm-dashboard/issues/94

### CODE REVIEW

**Comment by undera** [comment]
There is massive refactoring happens in #131 , will have to wait for it to complete

**Comment by undera** [comment]
Just to record my investigations, exerpt from Helm sources, `search_repo.go`:
```go
if o.devel { // search for releases and prereleases (alpha, beta, and release candidate releases).
		o.version = ">0.0.0-0"
	} else { // search only for stable releases, prerelease versions will be skip
		o.version = ">0.0.0"
	}
constraint, err := semver.NewConstraint(o.version)
v, err := semver.NewVersion(r.Chart.Version)
if constraint.Check(v) {
}
```

**Comment by undera** [comment]
The proper approach to this feature would include the checkbox in UI to enable display of development charts, passing that as parameter to API

**Comment by Bhargav-InfraCloud** [comment]
@undera Thanks for the pointers. It is working as expected. A query before submitting this PR:

How does the user enable development versions:
1. Directly from UI, the API response for the backend?
2. Starts the application using `--devel` which enables the option on UI to toggle on/off development versions, which is then sent to the backend as API response?

File: pkg/dashboard/objects/repos.go [hunk 1]
```diff
@@ -246,10 +249,23 @@ func (r *Repository) Charts() ([]*repo.ChartVersion, error) {
...
+
 	res := []*repo.ChartVersion{}
 	for _, v := range ind.Entries {
-		if len(v) > 0 { // TODO filter dev versions here, relates to #139
-			res = append(res, v[0])
+		if len(v) > 0 {
+			version, err := semver.NewVersion(v[0].Version)
+			if err != nil {
+				// Ignored if version string is not parsable
+				continue
```
Comment (undera) [inline pkg/dashboard/objects/repos.go:263]: I would log this situation with `Debugf`

**Comment by undera** [comment]
Two factors I take into account:
1. Seems that not too many users use devel versions in Helm
2. It would be harder to add a UI checkbox to install/upgrade modal for versions filter.

From that, I would do it as an environment variable. Once set, the user would get versions filtered accordingly.

Note that the filtering has to apply not only in `Charts()`, but also for `ByName()`

**Comment by Bhargav-InfraCloud** [comment]
@undera Please confirm the following:
1. In [Charts()][1], the current logic is only considering a single version (0'th index) of each chart. Any particular reason in doing so or should my changes iterate over all versions, filter, and append?
2. > From that, I would do it as an environment variable. Once set, the user would get versions filtered accordingly.

    If the `devel` is enabled/disabled only once from environment variables, why not have the `--devel` flag instead, as we initially discussed?


[1]: https://github.com/komodorio/helm-dashboard/blob/7a253350287c4216b4deb1c7cc2af441006f84d0/pkg/dashboard/objects/repos.go#L252

**Comment by undera** [comment]
1. The purpose of `Charts()` is to return the list of all charts, with the _latest_ version specified. So, if the chart only has dev versions, we don't display it when flag is not set. We can't return all the versions 'cause it will produce too much output.

2. You can add that flag. Just it will be harder to propagate it to all the objects. If that flag is rarely used, it might not be that important. Up to you here.

**Comment by Bhargav-InfraCloud** [comment]
@undera While I was trying to add tests for these changes, I noticed that they need interfaces for mocking. Otherwise, not testable.

However, when I started adding interfaces and tests, I realized that it is reorganizing most parts or `pkg/dashboard/objects`. So, I reverted those extra changes. I'll submit them under #210 if approved.

This card can only be tested manually. These are the sample calls I used to verify, with and without `--devel`. Please check and let me know.
```
curl http://localhost:8080/api/helm/repositories/kafka-operator | jq '.[].version'
curl http://localhost:8080/api/helm/repositories/versions\?name\=kafka-operator | jq '.[].version'
curl http://localhost:8080/api/helm/repositories/latestver\?name\=kafka-operator
```

File: go.mod [hunk 2]
```diff
@@ -30,7 +30,7 @@ require (
 	github.com/BurntSushi/toml v1.1.0 // indirect
 	github.com/MakeNowJust/heredoc v1.0.0 // indirect
 	github.com/Masterminds/goutils v1.1.1 // indirect
-	github.com/Masterminds/semver/v3 v3.1.1 // indirect
+	github.com/Masterminds/semver/v3 v3.1.1
```
Comment (undera) [inline go.mod:33]: If this dependency became explicit, please move it to the top section. This section is only for indirect dependencies.

**Comment by undera** [comment]
I have replied in #210 about testing. The code is testable, with the Helm's standard way of injecting fixtures. I don't see a real in in introducing interfaces.

**Comment by undera** [comment]
Do you plan to add the test @Bhargav-InfraCloud ?

**Comment by Bhargav-InfraCloud** [comment]
@undera My apologies, I didn't find time to update you on the progress. I've submitted the commit with test cases. Please review.

**Comment by undera** [comment]
Looks good. Awesome contributioN!

**Comment by Bhargav-InfraCloud** [comment]
Thanks, @undera. I appreciate the help throughout the discussion. :blush:

**Pull request merged**